### PR TITLE
Propose generic messages for commanding and getting states from controllers.

### PR DESCRIPTION
**Pull request by destogl**
I propose here some new messages for sending commands to controllers.
This work  was motivated by the implementation of the PID controller in ros2_controllers repository.

The message `MultiDOFCommand` should replace `std_msgs/MultiArrayFloat64` currently used with the controllers.
The former is deprecated, and a new message enables to set joint (or in general DOF) names, which makes it more explicit than just a list of values.

Other messages specify a generic status messages, providing flexibility about degrees of freedom.

The PR can also be separated if some messages are not so suitable.

### CODE REVIEW

File: control_msgs/msg/MultiDOFCommand.msg [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+# names of degrees of freedom
+string[] dof_names
+
+# values used by most of the controller
+float64[] values
+
+# First derivation of the values, e.g., velocity if values are positions.
+# This is useful for PID and similar controllers.
+float64[] values_dot
```
Comment (AndyZe) [inline control_msgs/msg/MultiDOFCommand.msg:13]: Just a suggestion

```suggestion
float64[] derivatives
```

File: control_msgs/msg/SingleDOFState.msg [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+string name
+
+# The set point, that is, desired state.
+float64 set_point
+
+# Current value of the process (ie: latest sensor measurement on the controlled value).
+float64 process_value
+
+# First time-derivative of the process value.
+float64 process_value_dot
```
Comment (AndyZe) [inline control_msgs/msg/SingleDOFState.msg:13]: Several other places to change `dot` to `derivative`, if you like that suggestion

File: control_msgs/msg/SingleDOFState.msg [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+# Current output of the controller.
+float64 command
+
+# Current PID parameters of the controller.
+float64 p
+float64 i
+float64 d
+float64 i_clamp
+bool antiwindup
```
Comment (AndyZe) [inline control_msgs/msg/SingleDOFState.msg:32]: I think `antiwindup` is redundant. If `i_clamp` is nonzero, then antiwindup will be applied, right?
```suggestion
float64 i_clamp
```

File: control_msgs/msg/MultiDOFCommand.msg (hunk 1 above)
Comment (AndyZe) [reply control_msgs/msg/MultiDOFCommand.msg:13]: I don't have a strong opinion. `dot` is easier to type but slightly harder to read

File: control_msgs/msg/SingleDOFState.msg [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+# First time-derivative of the error of the controlled value.
+float64 error_dot
+
+# Time between two consecutive updates/execution of the control law.
+float64 time_step
+
+# Current output of the controller.
+float64 command
+
+# Current PID parameters of the controller.
```
Comment (bmagyar) [inline control_msgs/msg/SingleDOFState.msg:27]: This is PID controller stuff in something called "SingleDOFState".  Nothing up to this point suggested that this message is specific for PID.

File: control_msgs/msg/SingleDOFState.msg
Comment (destogl) [inline control_msgs/msg/SingleDOFState.msg:1]: @christophfroehlich can you too check this file. Especially the  naming, what we discussed in ros-controls/roadmap#57.

File: control_msgs/msg/SingleDOFState.msg [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+float64 error
+
+# First time-derivative of the error of the controlled value.
+float64 error_dot
+
+# Time between two consecutive updates/execution of the control law.
+float64 time_step
+
+# Current output of the controller.
+float64 command
```
Comment (destogl) [inline control_msgs/msg/SingleDOFState.msg:25]: According to ros-controls/roadmap#57 the naming should be the following:

@christophfroehlich or do you like more the old naming (then we can adjust ros-controls/roadmap#57

```suggestion
float64 reference

# Current value of the process (ie: latest sensor measurement on the controlled value).
float64 feedback

# First time-derivative of the process value. E.g., velcoity.
float64 feedback_dot

# The error of the controlled value, essentially process_value - set_point (for a regular PID implementation).
float64 error

# First time-derivative of the error of the controlled value.
float64 error_dot

# Time between two consecutive updates/execution of the control law.
float64 time_step

# Current output of the controller.
float64 output
```

File: control_msgs/msg/SingleDOFStateStamped.msg
[the remark this answers is not in the document]
Comment (destogl) [reply control_msgs/msg/SingleDOFStateStamped.msg:1]: This is a good questions. I think we should remove those that are not used right now. We can add them as we need them.

[the remark this answers is not in the document]
Comment (christophfroehlich) [reply control_msgs/msg/SingleDOFStateStamped.msg:1]: I mean, one can also use the message with the vectors using only one entry for a single-dof..

File: control_msgs/msg/SingleDOFStateStamped.msg [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# This message presents current controller state of one degree of freedom.
+
+# Header timestamp should be update time of controller state
+std_msgs/Header header
+
+SingleDOFState dof_state
```
[the remark this answers is not in the document]
Comment (christophfroehlich) [reply control_msgs/msg/SingleDOFStateStamped.msg:6]: ok. is this used anywhere? I think at least it's not used within ros2_controllers.

[the remark this answers is not in the document]
Comment (christophfroehlich) [reply control_msgs/msg/SingleDOFStateStamped.msg:6]: shouldn't we deprecate/enhance [JointTrajectoryControllerState.msg](https://github.com/ros-controls/control_msgs/blob/3887cab9ff76f34c0d785ef7a8453e2367960ab6/control_msgs/msg/JointTrajectoryControllerState.msg) too?

[the remark this answers is not in the document]
Comment (destogl) [reply control_msgs/msg/SingleDOFStateStamped.msg:6]: I think we should do it in the next batch of PRs

**Review by bmagyar — APPROVED** [review]

**Comment by evan-palmer** [comment]
(I apologize in advance if this is the wrong place for this). Would it be possible to backport these messages to the humble branch?

**Comment by christophfroehlich** [comment]
I don't see any problems in backporting them.

**Comment by christophfroehlich** [comment]
@mergifyio backport humble

**Pull request merged**